tests/pessimistic/pipeline/pipeline.go

```diff
@@ -0,0 +1,329 @@
+	url := fmt.Sprintf("http://%s/schema/%s/%s", c.dbStatusAddr, db, table)
+	resp, err := httputil.NewHTTPClient(http.DefaultClient).Get(url)
+	if err != nil {
+		return
+	}
+	var body struct {
+		ID int64 `json:"id"`
+	}
+	err = json.Unmarshal(resp, &body)
+	if err != nil {
+		return
+	}
+	startKey, endKey = tablecodec.GetTableHandleKeyRange(body.ID)
+	startKey = codec.EncodeBytes([]byte{}, startKey)
+	endKey = codec.EncodeBytes([]byte{}, endKey)
+	return
+}
+
+// TearDown tears down the client.
+func (c *pipelineClient) TearDown(ctx context.Context, nodes []cluster.ClientNode, idx int) error {
+	c.db.ExecContext(ctx, `DROP DATABASE IF EXISTS `+dbName)
+	return c.db.Close()
+}
+
+// Invoke invokes a request to the database.
+// Mostly, the return Response should implement UnknownResponse interface
+func (c *pipelineClient) Invoke(ctx context.Context, node cluster.ClientNode, r interface{}) core.UnknownResponse {
+	panic("not implemented")
+}
+
+// NextRequest generates a request for latter Invoke.
+func (c *pipelineClient) NextRequest() interface{} {
+	panic("not implemented")
+}
+
+// DumpState the database state(also the model's state)
+func (c *pipelineClient) DumpState(ctx context.Context) (interface{}, error) {
+	panic("not implemented")
+}
+
+// Start runs self scheduled cases
+// this function will block Invoke trigger
+// if you want to schedule cases by yourself, use this function only
+func (c *pipelineClient) Start(ctx context.Context, cfg interface{}, clientNodes []cluster.ClientNode) error {
+	c.scatterTable(dbName, tableName)
+	defer func() {
+		c.checkSum(context.Background())
+		c.reportSummary()
+		c.stopScatterTable(dbName, tableName)
+	}()
+
+	for i := 0; i < c.TableSize; i++ {
+		for j := 0; j < c.ContenderCount; j++ {
+			go c.transfer(ctx, i)
+		}
+	}
+
+	checkTick := time.NewTicker(time.Second)
+	scheduleTick := time.NewTicker(c.ScheduleInterval)
+	reportTick := time.NewTicker(c.ReportInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-checkTick.C:
+			c.checkSum(ctx)
+
+		case <-scheduleTick.C:
+			c.schedule(ctx)
+
+		case <-reportTick.C:
+			c.reportStats()
+		}
+	}
+}
+
+func (c *pipelineClient) scatterTable(db, table string) error {
+	url := fmt.Sprintf("http://%s/tables/%s/%s/scatter", c.dbStatusAddr, db, table)
+	_, err := httputil.NewHTTPClient(http.DefaultClient).Get(url)
+	return err
+}
+
+func (c *pipelineClient) stopScatterTable(db, table string) error {
+	url := fmt.Sprintf("http://%s/tables/%s/%s/stop-scatter", c.dbStatusAddr, db, table)
+	_, err := httputil.NewHTTPClient(http.DefaultClient).Get(url)
+	return err
+}
+
+func (c *pipelineClient) checkKeyIsLocked(ctx context.Context, row int) (bool, error) {
+	tx, err := c.db.Begin()
+	if err != nil {
+		return false, err
+	}
+	defer tx.Rollback()
+	rows, err := tx.QueryContext(ctx, fmt.Sprintf(`SELECT * FROM %s WHERE id in (%d, %d) FOR UPDATE NOWAIT`, tableName, row, row+1))
```

Comment:
what does the `row + 1` means here? 
I think it should be `row + c.TableSize`?

Reply:
Yes. I missed it.